vimeo/vimeo-networking-java
# Adding new models for Live Video features

#### Ticket
[VA-2552](https://vimean.atlassian.net/browse/VA-2552)
[VA-2691](https://vimean.atlassian.net/browse/VA-2691)
[VA-2692](https://vimean.atlassian.net/browse/VA-2692)
[VA-2593](https://vimean.atlassian.net/browse/VA-2593)

#### Ticket Summary
Add support for the new data models and responses related to the live chat and live stats features.

#### Implementation Summary

- Added a new VimeoService request wrapper to get live stats.
- Added a live stats connection to the video object (this will require a new stats scope at the app level)
- Added a new video.live.chat object (LiveChat.java) that has the model data necessary to support the Firebase live chat feature for a live video as described here: [VA-2552](https://vimean.atlassian.net/browse/VA-2552)
- Modified the example projects so they will compile.
- Added a LiveStats object that represents the return value of a live stats request.
- Added new account types to the User object that have appeared in relation to the Live feature
- Added new is_staff, is_creator, and id to the User object. These are new fields that are needed for live chat display.
- Added access to the heartbeat uri that is available in video.play.(hls|dash).live.heartbeat
- Modified AppConfiguration.java (config) to use stag and added in a field for the new LiveConfiguration object which contains access to the heartbeat and chat config data. Also cleaned up getters/setters and member accessibility.
- Removed unused members from the FeaturesConfiguration object. There is currently a bug where none of these members are able to be initialized correctly due to the fact that the API is nesting these values within an iOS object. A ticket has been created with API to address this, but it is currently a moot point as the app isn't using any of these features in practice. 
- Added a LiveConfiguration object (represents config.live) to expose the LiveChatConfiguration and LiveHeartbeatConfiguration objects
- Added a LiveChatConfiguration object (represents config.live.chat) to contain the configuration for the live chat feature as described here: [VA-2552](https://vimean.atlassian.net/browse/VA-2552) 
- Added a LiveHeartbeatConfiguration object (represents config.live.heartbeat) to contain the configuration for the live chat feature as described here: [VA-2691](https://vimean.atlassian.net/browse/VA-2691) 

#### How to Test
There is a correspondingly named app level branch (VA-2691-Config-For-Stats) that is set up to utilize the new models. Confirm that it runs and behaves as expected. If you want, you can also add the following code in a place where you have access to a live video and use debug breakpoints to confirm that all of the expected data is available:

```
val roomId = video.live?.chat?.roomId
val token = video.live?.chat?.token
val user = video.live?.chat?.user
val userId = video.live?.chat?.user?.id
var userPictures = video.live?.chat?.user?.picturesList
var userName = video.live?.chat?.user?.name
var isVideoCreator = video.live?.chat?.user?.isVideoCreator

val hlsHeart = video.play?.hlsVideoFile?.liveHeartbeatUri
val dashHeart = video.play?.dashVideoFile?.liveHeartbeatUri
val heartConfig = ConfigurationManager.getInstance().mAppConfiguration?.liveHeartbeatConfiguration
val shouldHeart = heartConfig?.isEnabled ?: false
val heartInterval = heartConfig?.interval
val chatConfig = ConfigurationManager.getInstance().mAppConfiguration?.liveChatConfiguration

val appId = chatConfig?.appId
val apiKey = chatConfig?.apiKey
val authDomain = chatConfig?.authDomain
val databaseUrl = chatConfig?.databaseUrl
val messagingSenderId = chatConfig?.messagingSenderId
val projectId = chatConfig?.projectId
val storageBucket = chatConfig?.storageBucket

val liveStatsConnection = video.metadata?.connections?.liveStats?.uri ?: "farts"
VimeoClient.getInstance().getContent<LiveStats>(
	liveStatsConnection,
	CacheControl.FORCE_NETWORK,
	GetRequestCaller.LIVE_STATS,
	null,
	null,
	null,
	object : VimeoCallback<LiveStats>() {
	    override fun success(liveStats: LiveStats?) {
	        val currentViewerCount = liveStats?.currentViewerCount
	        val peakViewerCount = liveStats?.peakViewerCount
	        val plays = liveStats?.plays
	        val totalViewTime = liveStats?.totalViewTime
	    }

	    override fun failure(error: VimeoError?) {
	        TODO("not implemented")
	    }
	}
)

```
